gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/InstantiatingGrpcChannelProvider.java

```diff
@@ -410,6 +438,103 @@ ChannelCredentials createMtlsChannelCredentials() throws IOException, GeneralSec
+    if (trustBundle == null || privateKey == null || certChain == null) {
+      return null;
+    }
+    return TlsChannelCredentials.newBuilder()
+        .keyManager(privateKey, certChain)
+        .trustManager(trustBundle)
+        .build();
+  }
+
+  /**
+   * This method creates {@link ChannelCredentials} to be used by client to establish a plaintext
+   * connection to S2A. if {@param plaintextAddress} is not present, returns null.
+   *
+   * @param plaintextAddress the address to reach S2A which accepts plaintext connections
+   * @return {@link ChannelCredentials} to use to create a plaintext connection between client and
+   *     S2A
+   */
+  ChannelCredentials createPlaintextToS2AChannelCredentials(String plaintextAddress) {
+    if (Strings.isNullOrEmpty(plaintextAddress)) {
+      return null;
+    }
+    return S2AChannelCredentials.newBuilder(plaintextAddress, InsecureChannelCredentials.create())
+        .build();
+  }
+
+  /**
+   * This method creates gRPC {@link ChannelCredentials} configured to use S2A to estbalish a mTLS
+   * connection. First, the address of S2A is discovered by using the {@link S2A} utility to learn
+   * the {@code mtlsAddress} to reach S2A and the {@code plaintextAddress} to reach S2A. Prefer to
+   * use the {@code mtlsAddress} address to reach S2A if it is non-empty and the MTLS-MDS
+   * credentials can successfully be discovered and used to create {@link TlsChannelCredentials}. If
+   * there is any failure using mTLS-to-S2A, fallback to using a plaintext connection to S2A using
+   * the {@code plaintextAddress}. If {@code plaintextAddress} is not available, this function
+   * returns null; in this case S2A will not be used, and a TLS connection to the service will be
+   * established.
+   *
+   * @return {@link ChannelCredentials} configured to use S2A to create mTLS connection to
+   *     mtlsEndpoint.
+   */
+  ChannelCredentials createS2ASecuredChannelCredentials() {
+    S2A s2aUtils = S2A.newBuilder().build();
+    String plaintextAddress = s2aUtils.getPlaintextS2AAddress();
+    String mtlsAddress = s2aUtils.getMtlsS2AAddress();
+    if (Strings.isNullOrEmpty(mtlsAddress)) {
+      // Fallback to plaintext connection to S2A.
+      return createPlaintextToS2AChannelCredentials(plaintextAddress);
+    }
+    // Currently, MTLS to MDS is only available on GCE. See:
+    // https://cloud.google.com/compute/docs/metadata/overview#https-mds
+    // Try to load MTLS-MDS creds.
+    InputStream trustBundle = null;
+    InputStream privateKey = null;
+    InputStream certChain = null;
+    try {
+      trustBundle = new FileInputStream(MTLS_MDS_ROOT);
+      privateKey = new FileInputStream(MTLS_MDS_CERT_CHAIN_AND_KEY);
+      certChain = new FileInputStream(MTLS_MDS_CERT_CHAIN_AND_KEY);
+    } catch (FileNotFoundException ignore) {
+      // Fallback to plaintext-to-S2A connection.
+      LOG.log(
+          Level.INFO,
```

Comment:
Is this a misconfiguration or expected use case? If we are supposed to create a `S2ASecuredChannelCredentials` but failed due to `FileNotFoundException`, I would think it is a misconfiguration and should be a warning?

Reply:
This is an expected use case since currently MTLS-MDS is only GA on GCE. 

To be clear: the only time we fail to create `S2ASecuredChannelCredentials` is if both `mtlsAddress` and `plaintextAddress` are empty. 

When a `FileNotFoundException` happens, it means that there is some problem loading MTLS-MDS credentials, we still create `S2ASecuredChannelCredentials`, but we use plaintext-to-S2A connection, rather than a mtls-to-S2A connection. 

Because this exception can happen if MTLS-MDS is not available in that environment, I think we should leave it as INFO. WDYT?

Reply:
I see, if that's the case, we should check if the file exists, only create `MtlsToS2AChannelCredentials` if they all exist, and create `PlaintextToS2AChannelCredentials` if any of them does not exist. This is more intuitive and performant than using exceptions to determine logics. We don't have to add [null checks](https://github.com/googleapis/sdk-platform-java/pull/3326/files#diff-0ecb22815a443b8b1ae3712183cf63d5cd609fe326d28bc3d6cada228894a1d2R456) later as well.